### PR TITLE
feat(menu): color prop

### PR DESCRIPTION
**Pull request by bmdalex**
### Description

This PR:
- adds `color` prop to `Menu` component
- creates `Menu` color examples

### API

```jsx
<Menu color={COLOR} .../>
```
 where `COLOR` can be:
- a string: `'primary' | 'secondary' | 'blue' | 'green' | 'grey'
 | 'orange' | 'pink' | 'purple' | 'teal' | 'red' | 'yellow' | string`
- an object with:
  - key: one of `'foreground' | 'background' | 'border'` to identify the target area of the component
  - value: one of the strings above to describe  the color of the area
e.g.:
```jsx
<Menu color={ foreground: 'pink', background: 'yellow', border: 'green' }} .../>
```
renders:
screenshot 2019-01-14 at 02 47 47

### List of colored menu examples:
screenshot 2019-01-14 at 02 52 10

### CODE REVIEW

File: src/themes/types.ts [hunk 1]
```diff
@@ -99,12 +99,10 @@ export type ColorPalette = ExtendablePalette<
...
  * A type for the generic color scheme of a component based on CSS property names
  */
-export type ColorScheme = {
-  foreground: string
-  background: string
-  border: string
-  shadow: string
-}
+export type ColorScheme = Record<
+  'foreground' | 'background' | 'border' | 'shadow' | 'lighterBackground',
```
Comment (mnajdova) [inline src/themes/types.ts:103]: What is `lighterBackground` representing here? Maybe if we need different background it is result of some state inside the component. Should we consider adding object containing colors for the states: hovered, focus etc?

File: src/themes/teams/components/Menu/menuItemStyles.ts [hunk 2]
```diff
@@ -1,6 +1,13 @@
+import * as _ from 'lodash'
+
 import { getSideArrow } from '../../utils'
-import { pxToRem } from '../../../../lib'
-import { ComponentSlotStyleFunction, ComponentSlotStylesInput, ICSSInJSStyle } from '../../../types'
+import { pxToRem, getColorFromScheme } from '../../../../lib'
```
Comment (mnajdova) [inline src/themes/teams/components/Menu/menuItemStyles.ts:4]: I remember I used `getColorFromScheme` in the renderComponent, so that the user will receive colors object in the styles containing the merged colors. Please take a look on that. If that is not sufficient, we may rework the mechanism there and try to avoid this method in the styles.

File: src/themes/types.ts (hunk 1 above)
Comment (mnajdova) [reply src/themes/types.ts:103]: Or maybe it is a color of other type of the Menus?

File: src/lib/colorUtils.ts [hunk 3]
```diff
@@ -61,7 +62,7 @@ export const generateColorScheme = (
   // if the color prop is not defined, but the the color scheme is defined, then we are returning
   // the defaults from the color scheme if they exists
   if (colorScheme) {
-    return colorScheme && colorScheme.default ? colorScheme.default : {}
+    return colorScheme.default || {}
```
Comment (mnajdova) [inline src/lib/colorUtils.ts:65]: 👍

File: docs/src/examples/components/Menu/Variations/MenuExampleColor.shorthand.tsx [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+            color={{ foreground: colors[4], background: colors[6], border: colors[1] }}
+            items={items}
+          />
+        </Grid>
+      )
+    }}
+  />
+)
+
+export default MenuExampleColor
```
Comment (layershifter) [inline docs/src/examples/components/Menu/Variations/MenuExampleColor.shorthand.tsx:61]: This example looks overcomplicated to me, may be we can:
- introduce a new `Select` knob
- use it to get select an active color

File: docs/src/examples/components/Menu/Variations/MenuExampleVerticalPointingEnd.shorthand.tsx [hunk 5]
```diff
@@ -8,7 +8,7 @@ const items = [
 ]
 
 const MenuExampleVerticalPointingEnd = () => (
-  <Menu defaultActiveIndex={0} items={items} vertical pointing="end" />
+  <Menu defaultActiveIndex={0} items={items} vertical pointing="end" primary />
```
Comment (layershifter) [inline docs/src/examples/components/Menu/Variations/MenuExampleVerticalPointingEnd.shorthand.tsx:11]: This looks like a breaking change, is `primary` required now to get the same styling?

Comment (codepretty) [reply docs/src/examples/components/Menu/Variations/MenuExampleVerticalPointingEnd.shorthand.tsx:11]: Why not get rid of 'primary' altogether and just use the new color prop?

File: src/themes/teams/components/Menu/menuItemStyles.ts [hunk 6]
```diff
@@ -328,10 +368,13 @@ const menuItemStyles: ComponentSlotStylesInput<MenuItemPropsAndState, MenuVariab
 
         ...(primary
```
Comment (codepretty) [inline src/themes/teams/components/Menu/menuItemStyles.ts:369]: I think this would be greatly simplified (easier to redline & maintain) if we can remove `primary` prop and use just the `color` prop, since in Teams world `primary` is synonymous with just using brand color anyway. I think having `primary` prop makes sense on button and maybe some other components, but not for menu.

File: src/themes/base/colors.ts [hunk 7]
```diff
@@ -169,11 +169,13 @@ export const colorScheme: ColorSchemeMapping = _.mapValues(
       border: foreground,
       shadow: foreground,
       background: colorVariants[500],
+      lighterBackground: colorVariants[300],
```
Comment (codepretty) [inline src/themes/base/colors.ts:172]: What is the purpose of lighterBackground? The other keys make sense to me. Should lighterBackground really just be a different value of background?

File: docs/src/examples/components/Menu/Variations/MenuExampleVerticalPointingEnd.shorthand.tsx (hunk 5 above)
Comment (layershifter) [reply docs/src/examples/components/Menu/Variations/MenuExampleVerticalPointingEnd.shorthand.tsx:11]: Previously, we discussed this as I remember:
- replace `type` with `primary` and `secondary` props, #429
- remove `primary` and `secondary` from `Divider` as it doesn't applies any styling, #558
- on components like `Menu`, `primary` and `secondary` can affect styling, too: https://react.semantic-ui.com/collections/menu/#types-secondary

May be, in this case we should not expose `primary` and `secondary` as colors to avoid confusions.

File: src/themes/base/colors.ts (hunk 7 above)
Comment (bmdalex) [reply src/themes/base/colors.ts:172]: `lighterBackground` was just something temporary I came up before we came up with a more robust color palette that would allow different shades of background|foreground|border|etc color for different UI states (active|hovered|focused|etc).

**tl;dr**
will remove it soon and add something like:

```typescript
background: {
  initial: colorVariants[500],
  hover: colorVariants[300],
},
```

File: docs/src/examples/components/Menu/Variations/MenuExampleVerticalPointingEnd.shorthand.tsx (hunk 5 above)
Comment (bmdalex) [reply docs/src/examples/components/Menu/Variations/MenuExampleVerticalPointingEnd.shorthand.tsx:11]: @layershifter 
@codepretty 

We can agree there's a confusion from having `primary` and `color='primary'` props.
Let me just summarize our options:

### 1. remove `primary` color from `Menu` color palette and use `primary` prop

@layershifter I know we just discussed this, but after trying to code this up I realized how complicated it becomes because of the conditional logic for styles as @codepretty points out

We always need to check if the user set a color and if not default to either a primary variable color or a regular variable color (depending on `primary` prop value).

This cannot be extracted in a function because of all the possible shades of primary and regular colors.


### 2. change the logic for `primary` prop to be reflected by `color='primary'` instead (as @codepretty suggests); `primary` boolean prop does nothing for Teams theme

If we use `color='primary' as @codepretty suggests, the code will indeed be simplified tremendously as we do not have to handle the primary cases anymore.

Also, it looks like the colors are the only different styles for `<Menu primary .../>` so changing it to use `<Menu color='primary' .../> to achieve the purple menus does not cause other regressions!

To avoid confusion we should think about removing `primary` boolean prop or rename either the boolean prop or the name of the color?

What do others think?
@mnajdova @kuzhelov @miroslavstastny

File: src/themes/teams/components/Menu/menuItemStyles.ts (hunk 6 above)
Comment (bmdalex) [reply src/themes/teams/components/Menu/menuItemStyles.ts:369]: I agree, pls take a look at https://github.com/stardust-ui/react/pull/712#discussion_r247989391

File: docs/src/examples/components/Menu/Variations/MenuExampleColor.shorthand.tsx (hunk 4 above)
Comment (bmdalex) [reply docs/src/examples/components/Menu/Variations/MenuExampleColor.shorthand.tsx:61]: @layershifter can you show me an example of that? I would give it a shot

File: docs/src/examples/components/Menu/Variations/MenuExampleVerticalPointingEnd.shorthand.tsx (hunk 5 above)
Comment (codepretty) [reply docs/src/examples/components/Menu/Variations/MenuExampleVerticalPointingEnd.shorthand.tsx:11]: I talked with @levithomason yesterday about this and he also thinks the `primary`, `secondary` prop for Menu should be removed. I like the solution of just using `color` prop for all color style updates.

File: src/themes/base/colors.ts (hunk 7 above)
Comment (codepretty) [reply src/themes/base/colors.ts:172]: ah, ok, makes sense!

**Comment by bmdalex** [comment]
closing in favor of #756

File: src/themes/types.ts (hunk 1 above)
Comment (bmdalex) [reply src/themes/types.ts:103]: not relevant anymore, see #756

File: src/themes/teams/components/Menu/menuItemStyles.ts (hunk 2 above)
Comment (bmdalex) [reply src/themes/teams/components/Menu/menuItemStyles.ts:4]: not relevant anymore, see #756

File: docs/src/examples/components/Menu/Variations/MenuExampleVerticalPointingEnd.shorthand.tsx (hunk 5 above)
Comment (bmdalex) [reply docs/src/examples/components/Menu/Variations/MenuExampleVerticalPointingEnd.shorthand.tsx:11]: not relevant anymore, see #756
`primary` is not needed as a boolean prop any longer

File: src/themes/base/colors.ts (hunk 7 above)
Comment (bmdalex) [reply src/themes/base/colors.ts:172]: not relevant anymore, see #756
